ТЗ-15

IIExpanse opened the pull request.

IIExpanse committed:
Добавлен функционал по созданию запросов на добавление вещей, дописаны тесты.

IIExpanse committed:
Повышение покрытия..

IIExpanse committed:
Еще немного покрытия..

IIExpanse committed:
Еще немного покрытия..

IIExpanse committed:
Улучшено покрытие, дописан javadoc.

IIExpanse committed:
Возвращен на место удаленный маппер юзера

Nagash40 commented:
Андрей, добрый день.
Добавлен функционал с запросами и дописаны тесты. 
Из значимого в тестах - для репозиториев использована аннотация @DataJpaTest, чтобы убрать ненужную загрузку контекста. В остальном я использовал те же инструменты для написания интеграционных тестов, которые применял в прошлых ТЗ.

К объектам моделей и интерфейсам с пустой (и автогенерируемой) реализацией добавлена аннотация @Generated, чтобы преодолеть не самые адекватные требования о покрытии кода тестами (целевые уровни покрытия не менял).

andrew-larkin commented on src/test/java/ru/practicum/shareit/request/repository/ItemRequestRepositoryTest.java:
```diff
@@ -0,0 +1,72 @@
+        ItemRequest request1 = requestRepository.save(makeDefaultRequest(requester));
+        ItemRequest request2 = requestRepository.save(makeDefaultRequest(requester));
+
+        assertEquals(List.of(request2, request1),
```
падает вот этот assert

andrew-larkin commented on src/main/java/ru/practicum/shareit/booking/controller/BookingController.java:
```diff
@@ -1,58 +1,127 @@
 
-        return ResponseEntity.ok(bookingService.getBookingsByBookerIdOrOwnerIdAndStatusSortedByDateDesc(
-                bookerId, null, state));
+        if (size == null) {
```
всю логику нужно перенести в сервис, тут будет только вызов метода сервиса без доп.логики

andrew-larkin commented on src/main/java/ru/practicum/shareit/booking/controller/BookingController.java:
```diff
@@ -1,58 +1,127 @@
 
-        return ResponseEntity.ok(bookingService.getBookingsByBookerIdOrOwnerIdAndStatusSortedByDateDesc(
-                null, ownerId, state));
+        if (size == null) {
```
и здесь тоже

andrew-larkin commented on src/main/java/ru/practicum/shareit/booking/controller/BookingController.java:
```diff
@@ -1,58 +1,127 @@
+     * @throws UserNotFoundException - если пользователя с указанным id не существует.
+     */
     @GetMapping(path = "/owner")
     public ResponseEntity<Collection<BookingDto>> getBookingsByOwnerAndStatus(
```
чтобы не отправлять в ответе голую коллекцию, можно создать еще один dto (BookingListDto), куда положить эту коллекцию. json при этом не изменится, но визуально будет выглядеть менее нагруженно

andrew-larkin commented on src/main/java/ru/practicum/shareit/request/repository/ItemRequestRepository.java:
```diff
@@ -0,0 +1,17 @@
+
+@Repository
+@Generated
+public interface ItemRequestRepository extends JpaRepository<ItemRequest, Long> {
```
не увидел пагинации - стоит унаследоваться от PagingAndSortingRepository, передавать в запросах объект Pageable

andrew-larkin reviewed (changes requested):
Привет! Не хватает пагинации. В booking я оставил комментарии - они актуальны для всех контроллеров и сервисов. И в одном из тестов надо поправить assert.

IIExpanse committed:
fix: 1. Из контроллеров в сервисы убрана оставшаяся логика обработки параметров запросов.
2. Используются репозитории для работы с пагинацией, соответствующие изменения внесены в затронутый код.

Nagash40 commented:
Высылаю новую версию - вынес из контроллеров в сервисы лишнюю логику и поменял логику работы в пагинацией для использования репозиториев типа PagingAndSortingRepository.

Я не стал менять передачу коллекции на передачу коллекции в объекте, поскольку тесты постмана в текущем виде ожидают, что в теле ответа будет лежать голая коллекция. Проверка тела ответа начинается со строк вроде этой:
var jsonData = pm.response.json();
    pm.expect(jsonData.length, 'bookings count in total').to.eql(1);

Если передавать объект вместо коллекции, проект перестанет проходить постмановские тесты в GitHub Actions. Теоретически я могу создать новый класс и изменить правила его сериализации, чтобы на фронт уходила голая коллекция, но это лишает смысла упаковку коллекции в объект.
